base_layer/wallet/src/ffi.rs

```diff
@@ -29,87 +29,402 @@
 use chrono::NaiveDateTime;
-use std::os::raw::{c_char, c_int, c_uint, c_ulonglong};
-use tari_comms::connection::NetAddress;
+use libc::{c_char, c_int, c_uchar, c_uint, c_ulonglong};
+// use std::os::raw::{c_char, c_int, c_uint, c_ulonglong};
+use crate::{
+    output_manager_service::{handle::OutputManagerResponse::TransactionCancelled, OutputManagerConfig},
+    wallet::WalletConfig,
+};
+use std::{
+    boxed::Box,
+    ffi::{CStr, CString},
+    process::Output,
+    sync::{Arc, RwLock},
+    time::Duration,
+};
+use tari_comms::{
+    connection::NetAddress,
+    control_service::ControlServiceConfig,
+    peer_manager::{NodeIdentity, Peer, PeerFeature, PeerFeatures, PeerNodeIdentity},
+};
+use tari_comms_dht::envelope::NodeDestination::NodeId;
 use tari_core::{
-    transaction::{Transaction, UnblindedOutput},
+    tari_amount::MicroTari,
+    transaction::{
+        OutputFeatures,
+        OutputFlags,
+        Transaction,
+        TransactionInput,
+        TransactionKernel,
+        TransactionOutput,
+        UnblindedOutput,
+    },
     types::{PrivateKey, PublicKey},
 };
+use tari_crypto::keys::SecretKey;
+use tari_p2p::initialization::CommsConfig;
+use tari_utilities::{hex::Hex, ByteArray};
+use tokio::runtime::Runtime;
 
-// use tari_core::transaction_protocol::sender::RawTransactionInfo;
+pub type TariWallet = Wallet;
+pub type WalletDateTime = NaiveDateTime;
 
-// TODO implement this
-pub struct WalletFfi {}
+/// -------------------------------- ByteVector ------------------------------------------------ ///
+pub struct ByteVector(Vec<c_uchar>); // declared like this so that it can be exposed to external header
 
-pub struct IdentityFfi {
-    public_key: PublicKey,
-    secret_key: PrivateKey,
+#[no_mangle]
+pub unsafe extern "C" fn bytevector_create() -> *mut ByteVector {
+    let bytes = ByteVector(Vec::new());
+    Box::into_raw(Box::new(bytes))
 }
 
-pub struct KeyManagerStateFfi {
-    master_seed: PrivateKey,
-    branch_seed: String,
-    index: c_uint,
+#[no_mangle]
+pub unsafe extern "C" fn bytevector_destroy(bytes: *mut ByteVector) {
+    if bytes.is_null() {
+        let b = Box::from_raw(bytes);
+    }
+}
+
+/// adds c_char to internal vec
+pub unsafe extern "C" fn bytevector_add_byte(bytes: *mut ByteVector, byte: c_uchar) {
+    (*bytes).0.push(byte);
+}
+
+/// returns c_char at position in internal vector
+#[no_mangle]
+pub unsafe extern "C" fn bytevector_get_at(ptr: *mut ByteVector, i: c_int) -> c_uchar {
+    (*ptr).0.clone()[i as usize]
+}
+
+/// Returns the number of items, zero-indexed
+#[no_mangle]
+pub unsafe extern "C" fn bytevector_get_length(vec: *const ByteVector) -> c_int {
+    if vec.is_null() {
+        return 0;
+    }
+    (&*vec).0.len() as c_int
+}
+
+/// -------------------------------------------------------------------------------------------- ///
+
+/// -------------------------------- Public Key ------------------------------------------------ ///
+pub type WalletPrivateKey = PublicKey;
```

Comment:
You have aliased a PublicKey as a PrivateKey.

Rather than polluting the variable space with aliases you should rather just import this with an alias so it only applies to this file i.e. `use tari_core::types::PublicKey as WalletPublicKey`

Reply:
Why Alias it at all?

Reply:
My bad, switched them around in error, fixed in the next commit.

Alias is for the c header.